askup/views.py

```diff
@@ -1 +1,202 @@
+        """
+        context = super().get_context_data(**kwargs)
+        current_org = get_object_or_404(Qset, pk=self.kwargs.get('pk'))
+        context['main_title'] = current_org.name
+        context['current_qset_name'] = current_org.name
+        context['current_qset_id'] = current_org.id
+        context['is_admin'] = self.request.user.is_superuser
+        context['is_teacher'] = 'Teachers' in self.request.user.groups.values_list('name', flat=True)
+        context['is_student'] = 'Students' in self.request.user.groups.values_list('name', flat=True)
+        context['is_qset_creator'] = context['is_admin'] or context['is_teacher']
+        context['is_qset_allowed'] = True
+        return context
+
+    def get_queryset(self):
+        """
+        Get queryset for the list view.
+
+        Overriding the get_queryset of generic.ListView
+        """
+        pk = self.kwargs.get('pk')
+        user = self.request.user
+
+        if user.is_superuser:
+            return Qset.objects.filter(parent_qset_id=pk).order_by('name')
+        if user.is_authenticated():
+            return Qset.objects.filter(parent_qset=pk, top_qset__users=user.id).order_by('name')
+        else:
+            return []
+
+
+class QsetView(generic.ListView):
+    """Handles the Qset list view (subsets only/questions only/mixed)."""
+
+    model = Qset
+
+    def get_template_names(self):
+        """
+        Get template names to use in the view.
+
+        Overriding the get_template_names of generic.ListView
+        """
+        if self._current_qset_type == 1:
+            return ['askup/qset_subsets_only.html']
+        elif self._current_qset_type == 2:
+            return ['askup/qset_questions_only.html']
+        else:
+            return ['askup/qset_mixed.html']
+
+    def get_context_data(self, **kwargs):
+        """
+        Get context data for the list view.
+
+        Overriding the get_context_data of generic.ListView
+        """
+        current_qset = get_object_or_404(Qset, pk=self.kwargs.get('pk'))
+        self._current_qset_type = current_qset.type
+        context = super().get_context_data(**kwargs)
+
+        if self._current_qset_type == 2:
+            # Clear the qsets queryset if rendering the "questions only" Qset
+            context['object_list'] = []
+
+        if self._current_qset_type == 1:
+            # Clear the questions queryset if rendering the "qsets only" Qset
+            context['questions_list'] = []
+        else:
+            context['questions_list'] = Question.objects.filter(qset_id=self.kwargs.get('pk'))
+
+        context['main_title'] = current_qset.name
+        context['current_qset'] = current_qset
+        context['current_qset_name'] = current_qset.name
+        context['current_qset_id'] = current_qset.id
+        context['is_admin'] = self.request.user.is_superuser
+        context['is_teacher'] = 'Teachers' in self.request.user.groups.values_list('name', flat=True)
```

Comment:
I would move group_names out of the context fulfilling to avoid double database searching.
```
group_names = self.request.user.groups.values_list('name', flat=True)
context['is_teacher'] = 'Teachers' in group_names
context['is_student'] = 'Students' in group_names
```